### DISCUSSION TITLE
28.3

### DISCUSSION THREAD

**Discussion by njr0** [google/comprehensive-rust#1280] [General]
I guess I was expecting the code shown here to use `?` to simplify the error handling? Or was the idea that instructor or student did that.

Certainly, the first match seems suitable for using `?`:

```rust
    let mut username_file = username_file_result?;
```

The second one, because of the compatible types point, is slightly less straightforward. I changed it to:

```rust
    let _ = username_file.read_to_string(&mut username)?;
    Ok(username)
```
which seems to compile and run OK. (I presume it's OK because any error would have been propagated before the return value if it had occurred).

Not sure whether you intended to do the simplification, but it reads as if you did, either by showing both or just the simplified version.

**Comment by qwandor** [comment]
I think the idea here was that the instructor would demonstrate refactoring the code to use `?` in front of the class.

**Comment by njr0** [comment]
Sure. In all the cases before here, I think there's been a note to the instructor to this effect when they're supposed to do something.

(I'm just doing the course myself, so there is no instructor!)

**Reply by mgeisler** [reply]
Yeah, we should point this out in the speaker notes!
